File: cpp/day-09.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <ranges>
#include <unordered_map>
#include <unordered_set>
#include <sstream>
#include <utility>
#include <regex>
#include <stdint.h>
#include <tuple>
#include <stack>
#include <queue>
#include <functional>
#include <algorithm>
#include <optional>

class Block
{
public:
    Block() = delete;
    Block(size_t len) : _len(len) {}
    Block(size_t len, int64_t id) : _len(len), _id(id) {}

    auto len() const noexcept { return this->_len; }
    auto id() const noexcept { return this->_id.value(); }
    bool empty() const noexcept { return not this->_id.has_value(); }

private:
    size_t _len;
    std::optional<int64_t> _id;
};

auto part_one(const std::string &input) -> int64_t
{
    int64_t checksum = 0;
    std::vector<Block> disk;
    for (size_t i = 0; i < input.size(); i++)
    {
        const auto len = static_cast<size_t>(input[i] - '0');
        if (i % 2 == 0)
        {
            const auto id = static_cast<int64_t>(i / 2);
            for (size_t j = 0; j < len; j++)
            {
                disk.emplace_back(1, id);
            }
        }
        else
        {
            for (size_t j = 0; j < len; j++)
            {
                disk.emplace_back(1);
            }
        }
    }
    size_t first = 0;
    size_t last = disk.size() - 1;
    while (true)
    {
        while (first < disk.size() and not disk[first].empty())
            first++;
        while (0 < last && disk[last].empty())
            last--;
        if (last < first)
            break;
        std::swap(disk[first], disk[last]);
    }
    for (size_t i = 0; i < disk.size(); i++)
    {
        if (not disk[i].empty())
        {
            checksum += disk[i].id() * i;
        }
    }
    return checksum;
}

auto part_two(const std::string &input) -> int64_t
{
    int64_t checksum = 0;
    std::vector<Block> disk;
    for (size_t i = 0; i < input.size(); i++)
    {
        const auto len = static_cast<size_t>(input[i] - '0');
        if (len == 0)
            continue;
        if (i % 2 == 0)
        {
            const auto id = static_cast<int64_t>(i / 2);
            disk.emplace_back(len, id);
        }
        else
        {
            disk.emplace_back(len);
        }
    }
    size_t last = disk.size() - 1;
    while (0 < last)
    {
        size_t first = 0;
        while (0 < last && disk[last].empty())
            last--;
        while (first <= last)
        {
            if (disk[first].empty() and disk[last].len() <= disk[first].len())
                break;
            first++;
        }
        if (last < first)
        {
            last--;
            continue;
        }
        const auto len = static_cast<int64_t>(disk[first].len() - disk[last].len());
        disk[first] = disk[last];
        disk[last] = Block(disk[first].len());
        if (len > 0)
        {
            disk.emplace(disk.begin() + first + 1, len);
            last++;
        }
    }

    size_t index = 0;
    for (size_t i = 0; i < disk.size(); i++)
    {
        for (size_t j = 0; j < disk[i].len(); j++)
        {
            if (not disk[i].empty())
                checksum += disk[i].id() * index;
            index++;
        }
    }
    return checksum;
}

auto main() -> int
{

    std::string input;
    std::getline(std::cin, input);

    std::cout << "Part 1: " << part_one(input) << std::endl;
    std::cout << "Part 2: " << part_two(input) << std::endl;

    return 0;
}

File: rust/src/bin/day-09.rs
use std::collections::VecDeque;

#[derive(Debug, Clone)]
struct Block {
    len: usize,
    id: Option<i64>,
}

impl Block {
    fn new(len: usize, id: Option<i64>) -> Self {
        Block { len, id }
    }

    fn empty(&self) -> bool {
        self.id.is_none()
    }
}

fn part_one(input: &str) -> i64 {
    let mut checksum = 0;
    let mut disk: Vec<Block> = Vec::new();

    for (i, c) in input.chars().enumerate() {
        let len = (c.to_digit(10).unwrap_or(0)) as usize;
        if i % 2 == 0 {
            let id = (i / 2) as i64;
            for _ in 0..len {
                disk.push(Block::new(1, Some(id)));
            }
        } else {
            for _ in 0..len {
                disk.push(Block::new(1, None));
            }
        }
    }

    let mut first = 0;
    let mut last = disk.len().saturating_sub(1);

    while first < last {
        while first < disk.len() && !disk[first].empty() {
            first += 1;
        }
        while last > 0 && disk[last].empty() {
            last = last.saturating_sub(1);
        }

        if last <= first {
            break;
        }

        disk.swap(first, last);
    }

    for (i, block) in disk.iter().enumerate() {
        if let Some(id) = block.id {
            checksum += id * i as i64;
        }
    }

    checksum
}

fn part_two(input: &str) -> i64 {
    let mut checksum = 0;
    let mut disk: VecDeque<Block> = VecDeque::new();

    for (i, c) in input.chars().enumerate() {
        let len = (c.to_digit(10).unwrap_or(0)) as usize;
        if len == 0 {
            continue;
        }
        if i % 2 == 0 {
            let id = (i / 2) as i64;
            disk.push_back(Block::new(len, Some(id)));
        } else {
            disk.push_back(Block::new(len, None));
        }
    }
    let mut last = disk.len() - 1;
    while 0 < last {
        while 0 < last && disk[last].empty() {
            last -= 1;
        }
        let mut first = 0;
        while first <= last {
            if disk[first].empty() && disk[last].len <= disk[first].len {
                break;
            }
            first += 1;
        }

        if last < first {
            last -= 1;
            continue;
        }

        let last_block = disk[last].clone();
        let remaining_len = disk[first].len.saturating_sub(last_block.len);
        disk[first] = last_block;
        disk[last] = Block::new(disk[first].len, None);

        if remaining_len > 0 {
            disk.insert(first + 1, Block::new(remaining_len, None));
        }
    }

    let mut index = 0;
    for block in &disk {
        for _ in 0..block.len {
            if let Some(id) = block.id {
                checksum += id * index as i64;
            }
            index += 1;
        }
    }

    checksum
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let input = input.trim();

    println!("Part 1: {}", part_one(input));
    println!("Part 2: {}", part_two(input));
}
